subsys/net/lib/dhcpv4/dhcpv4_server.c

```diff
@@ -0,0 +1,1421 @@
+
+	if (slot < 0) {
+		LOG_ERR("No free DHCPv4 server intance.");
+		ret = -ENOMEM;
+		goto error;
+	}
+
+	ret = net_if_get_name(iface, ifreq.ifr_name, sizeof(ifreq.ifr_name));
+	if (ret < 0) {
+		LOG_ERR("Failed to obtain interface name.");
+		goto error;
+	}
+
+	sock = zsock_socket(AF_INET, SOCK_DGRAM, IPPROTO_UDP);
+	if (sock < 0) {
+		ret = errno;
+		LOG_ERR("Failed to create DHCPv4 server socket, %d", ret);
+		goto error;
+	}
+
+	ret = zsock_setsockopt(sock, SOL_SOCKET, SO_BINDTODEVICE, &ifreq,
+			       sizeof(ifreq));
+	if (ret < 0) {
+		ret = errno;
+		LOG_ERR("Failed to bind DHCPv4 server socket with interface, %d",
+			ret);
+		goto error;
+	}
+
+	ret = zsock_bind(sock, (struct sockaddr *)&addr, sizeof(addr));
+	if (ret < 0) {
+		ret = errno;
+		LOG_ERR("Failed to bind DHCPv4 server socket, %d", ret);
+		goto error;
+	}
+
+	fds[slot].fd = sock;
+	fds[slot].events = ZSOCK_POLLIN;
+
+	server_ctx[slot].iface = iface;
+	server_ctx[slot].sock = sock;
+	server_ctx[slot].server_addr = *server_addr;
+	server_ctx[slot].netmask = netmask;
+
+	k_work_init_delayable(&server_ctx[slot].timeout_work,
+			      dhcpv4_server_timeout);
+
+	LOG_DBG("Started DHCPv4 server, address pool:");
+	for (int i = 0; i < ARRAY_SIZE(server_ctx[slot].addr_pool); i++) {
+		server_ctx[slot].addr_pool[i].state = DHCPV4_SERVER_ADDR_FREE;
+		server_ctx[slot].addr_pool[i].addr.s_addr =
+					htonl(ntohl(base_addr->s_addr) + i);
+
+		LOG_DBG("\t%2d: %s", i,
+			net_sprint_ipv4_addr(
+				&server_ctx[slot].addr_pool[i].addr));
+	}
+
+	ret = net_socket_service_register(&dhcpv4_server, fds, ARRAY_SIZE(fds),
+					  NULL);
+	if (ret < 0) {
+		LOG_ERR("Failed to register socket service, %d", ret);
+		memset(&server_ctx[slot], 0, sizeof(server_ctx[slot]));
+		fds[slot].fd = -1;
+		goto error;
+	}
+
+	k_mutex_unlock(&server_lock);
+
+	return 0;
+
+error:
+	if (sock >= 0) {
+		(void)zsock_close(sock);
+	}
+
+	k_mutex_unlock(&server_lock);
+
+	return ret;
+}
+
+int net_dhcpv4_server_stop(struct net_if *iface)
+{
+	struct k_work_sync sync;
+	int slot = -1;
+	int ret = 0;
+	bool service_stop = true;
+
+	if (iface == NULL) {
+		return -EINVAL;
+	}
+
+	k_mutex_lock(&server_lock, K_FOREVER);
+
+	for (int i = 0; i < ARRAY_SIZE(server_ctx); i++) {
+		if (server_ctx[i].iface == iface) {
+			slot = i;
+			break;
+		}
+	}
+
+	if (slot < 0) {
+		ret = -ENOENT;
+		goto out;
+	}
+
+	fds[slot].fd = -1;
+	(void)zsock_close(server_ctx[slot].sock);
+
+	k_work_cancel_delayable_sync(&server_ctx[slot].timeout_work, &sync);
+
+	memset(&server_ctx[slot], 0, sizeof(server_ctx[slot]));
+
+	for (int i = 0; i < ARRAY_SIZE(fds); i++) {
+		if (fds[i].fd >= 0) {
+			service_stop = false;
+			break;
+		}
+	}
+
+	if (service_stop) {
+		ret = net_socket_service_unregister(&dhcpv4_server);
+	} else {
+		ret = net_socket_service_register(&dhcpv4_server, fds,
+						  ARRAY_SIZE(fds), NULL);
+	}
+
+out:
+	k_mutex_unlock(&server_lock);
+
+	return ret;
+}
+
+int net_dhcpv4_server_foreach_lease(struct net_if *iface,
```

Comment:
Just wondering should there be a way to go through all the leases over all interfaces, for example user could set the iface to NULL to achieve this?

Reply:
The function now accepts NULL iface, and loops over all active server instances.